### PR TITLE
rgw/s3website: ErrorDoc & Double-fault support

### PR DESCRIPTION
**Pull request by robbat2**

### CODE REVIEW

**Comment by robbat2** [comment]
With this PR and PR#8633, the s3-tests now include a full 100% pass on the new s3website functionality.
https://github.com/ceph/s3-tests/pull/92

**Comment by robbat2** [comment]
Review-copy of this patch, minus the whitespace changes (a block of abort_early is indented as conditional now)
http://dev.gentoo.org/~robbat2/ceph-fcb3cf169f1f.diff

**Comment by robbat2** [comment]
GitHub view with whitespace ignored (w=1 urlparam):
https://github.com/ceph/ceph/pull/8669/files?w=1

File: src/rgw/rgw_rest.cc [hunk 1]
```diff
@@ -699,49 +695,56 @@ void abort_early(struct req_state *s, RGWOp *op, int err_no,
...
-      /*
-       * reqest_uri is always start with slash, so we need to remove
-       * the unnecessary slash at the end of dest_uri.
-       */
-      if (dest_uri[dest_uri.size() - 1] == '/') {
-        dest_uri = dest_uri.substr(0, dest_uri.size() - 1);
+
+  // If the error handler(s) above dealt with it completely, they should have
+  // returned 0. If non-zero, we need to continue here.
+  if(err_no) {
```
Comment (yehudasa) [inline src/rgw/rgw_rest.cc]: @robbat2 missing space after if

File: src/rgw/rgw_rest.cc [hunk 2]
```diff
@@ -699,49 +695,56 @@ void abort_early(struct req_state *s, RGWOp *op, int err_no,
...
+    // Watch out, we might have a custom error state already set!
+    if(s->err.http_ret && s->err.http_ret != 200) {
```
Comment (yehudasa) [inline src/rgw/rgw_rest.cc]: and here too, and probably in some other places

File: src/rgw/rgw_rest_s3.cc [hunk 3]
```diff
@@ -3951,52 +3956,87 @@ RGWOp* RGWHandler_REST_S3Website::op_head()
...
+  getop->init(store, s, this);
+  getop->range_str = NULL;
+  getop->if_mod = NULL;
+  getop->if_unmod = NULL;
+  getop->if_match = NULL;
+  getop->if_nomatch = NULL;
+  s->object = errordoc_key;
+
+  ret = init_permissions(getop);
+  if (ret < 0) {
```
Comment (yehudasa) [inline src/rgw/rgw_rest_s3.cc]: @robbat2 `getop->put()`? unless something drops reference at some higher level, maybe use intrusive_ptr?

File: src/rgw/rgw_rest_s3.cc [hunk 4]
```diff
@@ -3951,52 +3956,87 @@ RGWOp* RGWHandler_REST_S3Website::op_head()
...
+  ret = getop->verify_params();
+  if(ret < 0) {
+    ldout(s->cct, 20) << "serve_errordoc failed, verify_params ret=" << ret << dendl;
+    return -1; // Trigger double error handler
+  }
+
+  // No going back now
+  getop->pre_exec();
+  /*
+   * FIXME Missing headers:
```
Comment (yehudasa) [inline src/rgw/rgw_rest_s3.cc]: maybe open a ceph tracker issue for this FIXME?

**Comment by yehudasa** [comment]
@robbat2 see my comments. overall fine, a few minor issues.

File: src/rgw/rgw_rest_s3.cc (hunk 3 above)
Comment (robbat2) [inline src/rgw/rgw_rest_s3.cc]: Clarifying @yehudasa's note for anybody else reading this:
`put_op(getop)`; needed to cleanup at the end of the function.
Alternatively (better route) use `shared_ptr` to GC `getop`.

**Comment by robbat2** [comment]
@yehudasa the two new commits fix the issue you pointed out. Not squashed together for ease of updated review since you reviewed the prior commits already.

**Pull request merged**